### PR TITLE
support dump specified tables

### PR DESCRIPTION
**Pull request by wusendong**
Implement the issue #2

### CODE REVIEW

**Comment by wusendong** [comment]
I already signed the google CLA, could you verify again? @googlebot I signed it!

File: dump.go [hunk 1]
```diff
@@ -110,12 +119,31 @@ func (d *Dumper) DumpDDLs(ctx context.Context) error {
...
+		if len(d.tables) > 0 && !d.tables[parsesTableName(ddl)] {
+			continue
+		}
 		fmt.Fprintf(d.out, "%s;\n", ddl)
 	}
 
 	return nil
 }
 
+func parsesTableName(statement string) string {
```
Comment (yfuruyama) [inline dump.go]: nit: `s/parsesTableName/parseTableName/` as we don't use third-person singular verb for function name.

Comment (yfuruyama) [inline dump.go]: nit: consider to rename `statement` to `ddl` as it makes clearer that we parse table name for DDL.

File: dump.go [hunk 2]
```diff
@@ -110,12 +119,31 @@ func (d *Dumper) DumpDDLs(ctx context.Context) error {
...
+		return match[3]
+	}
+	if tableRegexp.MatchString(statement) {
+		match := tableRegexp.FindStringSubmatch(statement)
+		return match[1]
+	}
+	return ""
+}
+
+var indexRegexp = regexp.MustCompile(`^CREATE (UNIQUE |NULL_FILTERED )?INDEX ([a-zA-Z0-9_]+) ON ([a-zA-Z0-9_]+)`)
```
Comment (yfuruyama) [inline dump.go]: 1. Can we use `\s+` for whitespace? It is not guaranteed that DDL has only single whitespace between keywords.
1. Please use uncaptured grouping `(?:)` for unnecessary grouping.
1. If index_name or table_name has [reserved words](https://cloud.google.com/spanner/docs/data-definition-language#reserved_keywords), name is escaped with backticks "`". Please consider taht.
```
CREATE INDEX `order` ON TABLE(`by`)
```

File: dump_test.go [hunk 3]
```diff
@@ -0,0 +1,44 @@
+package main
+
+import "testing"
+
+func Test_parsesTableName(t *testing.T) {
```
Comment (yfuruyama) [inline dump_test.go]: nit: use capital letter after "Test", `TestParseTableName`

File: dump_test.go [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+	tests := []struct {
```
Comment (yfuruyama) [inline dump_test.go:6]: Please add tests for escaped identifier.

File: main.go [hunk 5]
```diff
@@ -36,7 +37,8 @@ type options struct {
 
 func main() {
 	var opts options
-	if _, err := flags.Parse(&opts); err != nil {
+	tables, err := flags.Parse(&opts)
```
Comment (yfuruyama) [inline main.go]: Instead of command line argument, can we use command line option to specify table_name like `--tables=t1,t2,...`?

Unlike mysqldump, this tool uses options to get even required parameters.

File: README.md [hunk 6]
```diff
@@ -7,7 +7,7 @@ Exported databases can be imported to Cloud Spanner with [spanner-cli](https://g
 
 ```sh
 # Export
-$ spanner-dump -p ${PROJECT} -i ${INSTANCE} -d ${DATABASE} > data.sql
+$ spanner-dump -p ${PROJECT} -i ${INSTANCE} -d ${DATABASE} [table_name ...] > data.sql
```
Comment (yfuruyama) [inline README.md]: Please revert this change.

File: README.md [hunk 7]
```diff
@@ -40,7 +40,7 @@ go get -u github.com/cloudspannerecosystem/spanner-dump
 
 ```
 Usage:
-  spanner-dump [OPTIONS]
+  spanner-dump [OPTIONS] [table_name ...]
```
Comment (yfuruyama) [inline README.md]: Please revert this change.

File: main.go [hunk 8]
```diff
@@ -20,23 +20,27 @@ package main
...
 	"time"
+
+	"github.com/jessevdk/go-flags"
 )
 
 type options struct {
 	ProjectId  string `short:"p" long:"project" description:"(required) GCP Project ID."`
 	InstanceId string `short:"i" long:"instance" description:"(required) Cloud Spanner Instance ID."`
 	DatabaseId string `short:"d" long:"database" description:"(required) Cloud Spanner Database ID."`
+	Tables     string `short:"t" long:"tables" description:"comma-separated table names, e.g. \"table1,table2\" "`
```
Comment (yfuruyama) [inline main.go]: Can we remove short name `"t"` from the option? I want to reserve the short name for future use.

File: main.go [hunk 9]
```diff
@@ -20,23 +20,27 @@ package main
...
 	NoDDL      bool   `long:"no-ddl" description:"No DDL information."`
 	Timestamp  string `long:"timestamp" description:"Timestamp for database snapshot in the RFC 3339 format."`
 	BulkSize   uint   `long:"bulk-size" description:"Bulk size for values in a single INSERT statement."`
 }
 
 func main() {
 	var opts options
-	if _, err := flags.Parse(&opts); err != nil {
+	_, err := flags.Parse(&opts)
```
Comment (yfuruyama) [inline main.go]: nit: please define variables in if block to narrow the scope.
```
if _, err := flags.Parse(&opts); err != nil { ...
```

**Comment by wusendong** [comment]
Thanks for your kindly review~ your suggestion is very useful for me

File: main.go [hunk 10]
```diff
@@ -54,7 +58,8 @@ func main() {
 	}
 
 	ctx := context.Background()
-	dumper, err := NewDumper(ctx, opts.ProjectId, opts.InstanceId, opts.DatabaseId, os.Stdout, timestamp, opts.BulkSize)
+	tables := strings.Split(opts.Tables, ",")
```
Comment (yfuruyama) [inline main.go]: If opts.Tables is empty string, `strings.Split(opts.Tables, ",")` produces a slice which has a single value `""`.
https://play.golang.org/p/gaOXKS58kw0

This should be an empty slice, otherwise spanner-dump can't work without `--tables` option.

**Review by yfuruyama — CHANGES_REQUESTED** [review]

File: main.go (hunk 10 above)
Comment (wusendong) [reply main.go]: you are right, let me change it

File: dump.go [hunk 11]
```diff
@@ -110,12 +119,31 @@ func (d *Dumper) DumpDDLs(ctx context.Context) error {
...
+	if tableRegexp.MatchString(ddl) {
+		match := tableRegexp.FindStringSubmatch(ddl)
+		return match[1]
+	}
+	return ""
+}
+
+var indexRegexp = regexp.MustCompile("^\\s*CREATE\\s+(?:UNIQUE\\s+|NULL_FILTERED\\s+)?INDEX\\s+(?:[a-zA-Z0-9_`]+)\\s+ON\\s+`?([a-zA-Z0-9_]+)`?")
+var tableRegexp = regexp.MustCompile("^\\s*CREATE\\s+TABLE\\s+`?([a-zA-Z0-9_]+)`?")
+
```
Comment (yfuruyama) [inline dump.go:150]: I'm sorry to bother you many times, but I found DDL could have `ALTER TABLE table_name` statement when we use [Foreign Key](https://cloud.google.com/spanner/docs/foreign-keys/how-to) constraints.

For example, This is what I get from Spanner for DDLs containing foreign key.
```
CREATE TABLE t5 (
  Id INT64 NOT NULL,
  T6Id INT64 NOT NULL,
) PRIMARY KEY(Id);

CREATE TABLE t6 (
  Id INT64 NOT NULL,
) PRIMARY KEY(Id);

ALTER TABLE t5 ADD FOREIGN KEY(T6Id) REFERENCES t6(Id);
```

Could you consider supporting this as well?

**Review by yfuruyama — CHANGES_REQUESTED** [review]

File: dump.go (hunk 11 above)
Comment (wusendong) [reply dump.go:150]: No problem, let me add foreign key support

**Review by yfuruyama — APPROVED** [review]
@wusendong LGTM. Thanks for contributing this!
